controllers/growth.js

```diff
@@ -3,45 +3,208 @@
+        var simple_key;
+
+        if (data.month) {
+            end_month = moment(data.month + '-01').format('YYYY-MM');
+        }
+
+        if (+data.duration) {
+            start_month = moment(end_month + '-01')
+                .subtract(+data.duration, 'month').format('YYYY-MM');
+        }
+
+        console.log('from ' + start_month + ' to ' + end_month);
+        console.log(data.network);
+
+        simple_key = data.network.split(',') + ':' + start_month + ':' + end_month;
+        cache_key = crypto.createHash('md5').update(simple_key).digest('hex');
 
-            mysql.open(config.DB)
-                .query(
-                    query, [data.id, from, to],
-                    send_response
+        // check cache first
+        cache.get(cache_key, retrieve_cache);
+    }
+
+    function retrieve_cache (err, cache_result) {
+
+        if (!err || !cache_result) {
+
+            // retrieve list of channels
+            return mysql.open(config.DB)
+                .query('SELECT SQL_CACHE channel_id, title, linked_date '
+                       + 'FROM mcn_channels '
+                       + 'WHERE network IN (?) '
+                       + 'AND DATE_FORMAT(linked_date, "%Y-%m") = ?',
+                       [data.network.split(','), start_month],
+                       get_stat
                 )
                 .end();
-        },
+        }
+
+        return res.send(cache_result);
+    }
+
+    function get_stat (err, result) {
+        var channel_ids = [];
+
+        if (err) {
+            return next(err);
+        }
+
+        if (!result.length) {
+            return res.send({ results: [] });
+        }
+
+        channels = result;
+
+        channels.forEach(function (row) {
+            channel_ids.push(row.channel_id);
+        });
+
+        console.log('no of channels', channel_ids.length);
+
+        mysql.open(config.DB)
+            .query(
+                'SELECT SQL_CACHE DATE_FORMAT(insert_date, "%Y-%m") as month, '
+                + 'channel_id, MAX(views) as views, '
+                + 'MAX(subscribers) as subscribers, MAX(comments) as comments '
+                + 'FROM channel_stats WHERE channel_id IN (?) '
+                + 'AND DATE_FORMAT(insert_date, "%Y-%m") BETWEEN ? AND ? '
+                + 'GROUP BY month, channel_id '
+                + 'ORDER BY month',
+                [channel_ids, start_month, end_month],
+                send_response
+            )
+            .end();
+    }
+
+    function send_response (err, result) {
+        var count = 0,
+            new_date = 0,
+            response = { results: [], channels: result };
+
+        if (err) {
+            return next(err);
+        }
+
+        if (!result.length) {
+            return res.send({ results: [] });
+        }
+
+        result.forEach(function (row) {
+            if (count) {
+                if (row.month !== result[count-1].month) {
+                    response.results.push(result[new_date]);
```

Comment:
use another variable instead of `result`, `result` is the data returned by query